### PR TITLE
Infer `::from()` and `::tryFrom()` return types on backed enums

### PR DESCRIPTION
**Pull request by weirdan**
Fixes vimeo/psalm#6429

### CODE REVIEW

File: src/Psalm/Internal/Codebase/Methods.php [hunk 1]
```diff
@@ -686,21 +694,55 @@ public function getMethodReturnType(
...
+                || $original_method_name === 'tryfrom'
+                ) && $source_analyzer
+                && isset($args[0])
+                && ($first_arg_type = $source_analyzer->getNodeTypeProvider()->getType($args[0]->value))
+            ) {
+                $types = [];
+                foreach ($original_class_storage->enum_cases as $case_name => $case_storage) {
+                    if (UnionTypeComparator::isContainedBy(
+                        $source_analyzer->getCodebase(),
+                        \is_int($case_storage->value) ?
```
Comment (orklah) [inline src/Psalm/Internal/Codebase/Methods.php:730]: Is it possible to use a constant to initialize a case? If it happens, is_int() will fail. Can't we access the type at this moment?

Comment (weirdan) [reply src/Psalm/Internal/Codebase/Methods.php:730]: It's possible, and it fails. But not here.

https://3v4l.org/WHXIV#v8.1rc3
https://psalm.dev/r/3547ca91d3

Comment (weirdan) [reply src/Psalm/Internal/Codebase/Methods.php:730]: Built-in constants work. 
User-defined constants don't: https://3v4l.org/nhGRA#v8.1rc3
Class constants work if the class is defined before the enum: https://3v4l.org/bcYPN#v8.1rc3.
Class constants from autoloaded classes do not work: https://3v4l.org/AjOCi#v8.1rc3

3v4l doesn't have proper 8.1 yet, so it's based on RC3. Some of the above could have been fixed before final release.

Comment (weirdan) [reply src/Psalm/Internal/Codebase/Methods.php:730]: > Some of the above could have been fixed before final release.

I've just built 8.1 branch, and there are no changes for those cases.

**Pull request merged**